SI-5314 - CPS transform of return statement fails (resubmission of #987)

phaller opened the pull request:
Enable return expressions in CPS code if they are in tail position. Note that tail returns are
only removed in methods that do not call `shift` or `reset` (otherwise, an error is reported).

Addresses the issues pointed out in a previous pull request:
https://github.com/scala/scala/pull/720
- Addresses all issues mentioned here:
  https://github.com/scala/scala/pull/720#issuecomment-6429705
- Move transformation methods to SelectiveANFTransform.scala:
  https://github.com/scala/scala/pull/720#commitcomment-1477497
- Do not keep a list of tail returns.

Tests:
- continuations-neg/t5314-missing-result-type.scala
- continuations-neg/t5314-type-error.scala
- continuations-neg/t5314-npe.scala
- continuations-neg/t5314-return-reset.scala
- continuations-run/t5314.scala
- continuations-run/t5314-2.scala
- continuations-run/t5314-3.scala

Review by @TiarkRompf and @adriaanm

phaller committed:
SI-5314 - CPS transform of return statement fails

Enable return expressions in CPS code if they are in tail position. Note that tail returns are
only removed in methods that do not call `shift` or `reset` (otherwise, an error is reported).

Addresses the issues pointed out in a previous pull request:
https://github.com/scala/scala/pull/720

- Addresses all issues mentioned here:
  https://github.com/scala/scala/pull/720#issuecomment-6429705

- Move transformation methods to SelectiveANFTransform.scala:
  https://github.com/scala/scala/pull/720#commitcomment-1477497

- Do not keep a list of tail returns.

Tests:
- continuations-neg/t5314-missing-result-type.scala
- continuations-neg/t5314-type-error.scala
- continuations-neg/t5314-npe.scala
- continuations-neg/t5314-return-reset.scala
- continuations-run/t5314.scala
- continuations-run/t5314-2.scala
- continuations-run/t5314-3.scala

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/676/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/676/

TiarkRompf commented on src/continuations/plugin/scala/tools/selectivecps/CPSAnnotationChecker.scala:
```diff
@@ -466,6 +480,13 @@ abstract class CPSAnnotationChecker extends CPSUtils with Modes {
           tpe
 
+        case ret @ Return(expr) =>
+          // only change type if this return will (a) be removed (in tail position) or (b) cause
```
I still feel uneasy about changing the types here and I don't feel confident that I understand how the condition in the code (hasPlusMarker) corresponds exactly to the condition that's stated in the comment.

phaller replied:
In commit 834a8faa123f168e8baea772b06ebce2874ff431 I moved the adaptation of types of return expressions from `addAnnotations` to `typedReturn` via the `adaptTypeOfReturn` hook, which is a new hook I added to `AnnotationCheckers`.

In commit cd14c3087ceca8145c3013c2751648b936f1256d I also improved the doc comment on `adaptTypeOfReturn` of `CPSAnnotationCheckers` to explain exactly when the type of a return is adapted:

```
/** Returns an adapted type for a return expression if the method's result type (pt) is a CPS type.
 *  Otherwise, it returns the `default` type (`typedReturn` passes `NothingClass.tpe`).
 *  
 *  A return expression in a method that has a CPS result type is an error unless the return
 *  is in tail position. Therefore, we are making sure that only the types of return expressions
 *  are adapted which will either be removed, or lead to an error.
 */
```

TiarkRompf commented on src/continuations/plugin/scala/tools/selectivecps/CPSUtils.scala:
```diff
@@ -45,10 +47,15 @@ trait CPSUtils {
   lazy val allCPSAnnotations = List(MarkerCPSSym, MarkerCPSTypes, MarkerCPSSynth,
     MarkerCPSAdaptPlus, MarkerCPSAdaptMinus)
 
+  lazy val allCPSMethods = List(MethShiftUnit, MethShiftUnit0, MethShiftUnitR, MethShift, MethShiftR,
```
for what measure of 'all'? any method that calls shift can be considered a cps method ...

phaller replied:
Removed since no longer needed

TiarkRompf commented on test/files/continuations-neg/t5314-return-reset.scala:
```diff
@@ -0,0 +1,21 @@
+  def caller(): Int = {
+    val v: Int = reset {
+      val res: Int = bar(8)
+      if (rnd.nextInt(100) > 50) return 5 // not allowed, since method is calling `reset`
```
What happens if I define `def myInnocentMethod[T,U](block: => T@cps[U]) = reset(block)` and use that instead of reset?

phaller replied:
This is now fixed by replacing the `CheckCPSMethodsTraverser`

TiarkRompf commented on src/continuations/plugin/scala/tools/selectivecps/SelectiveANFTransform.scala:
```diff
@@ -32,6 +32,84 @@ abstract class SelectiveANFTransform extends PluginComponent with Transform with
+    /* Does not attempt to remove tail returns.
+     * Only checks whether the method contains returns as well as calls to CPS methods.
+     */
+    class CheckCPSMethodsTraverser extends Traverser {
```
This is way to much complexity for my taste, the previous version was much easier to follow. Also, what is the allCPSMethods check good for? It will only catch literal uses of shift, reset etc but no transitive ones like `future { ... }`.

TiarkRompf replied:
Maybe an alternative approach would be to make the Return case in transValue inspect the cpsR parameter to decide whether to flag an error (cpsR should carry information whether the current block is cps or not).

phaller replied:
That's an interesting idea. In the example below (t5314-return-reset.scala), both `cpsR` and `cpsA` are `None`, though, so no error would be flagged. Should that case be OK?

TiarkRompf replied:
Good point, cpsR only describes the statically innermost expected context type, which is pure in this case. So it seems one option would be to turn it into a list of parent context types, so that the Return case can check if any parent is impure. Another option would be to add a Boolean parameter isAnyParentImpure to the traversal methods to maintain this information directly.

phaller replied:
Yes, I think that's a good idea. Yesterday evening I came to the same conclusion, but with a different implementation. To avoid an additional traversal I will pursue the approach with the `isAnyParentImpure` parameter first.

phaller replied:
OK, I replaced the `CheckCPSMethodsTraverser` with an additional `isAnyParentImpure` parameter on the transformer methods (`transExpr` and friends). This is definitely simpler and it resolves some of the other issues that you pointed out.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/10/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/10/

phaller committed:
Replace CheckCPSMethodTraverser with additional parameter on transformer methods

Other fixes:

- remove CPSUtils.allCPSMethods
- add clarifying comment about adding a plus marker to a return expression

phaller commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/129/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/833/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/129/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/833/

phaller committed:
Add missing cases in tail return transform

Disabled warnings that no longer apply because of tail returns.
Add several test cases.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/149/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/854/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/149/

phaller committed:
Revert "Add missing cases in tail return transform"

This reverts commit 8d020fab9758ced93eb18fa51c906b95ec104aed.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/854/

adriaanm commented:
@TiarkRompf could you have a look and sign off if it looks good to you, please?

phaller commented:
@TiarkRompf and myself discovered an irregularity that I'm currently trying to resolve. I will notify everyone as soon as I have a resolution.

phaller commented:
I just made a break through on this! I will clean up my working copy, and then submit hopefully the last update.

phaller committed:
Simplify the adaptation of types of return expressions

Add `adaptTypeOfReturn` hook to `AnnotationCheckers`.
Move adaptation of types of return expressions from `addAnnotations`
to `typedReturn` via `adaptTypeOfReturn` hook.

This resolves an inconsistency where previously types could have
a plus marker without additional CPS annotations. This also adds
additional test cases.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/184/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/889/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/184/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/889/

adriaanm commented:
great! now all this needs is a fiat (of the lower-case kind)

paulp commented:
This needs to get into 2.10.x if it's going to go into 2.10.  We are still waiting for someone to approve this right?

lrytz commented:
@phaller please comment, and ping the reviewers if necessary.

phaller committed:
Improve doc comment on adaptTypeOfReturn in CPSAnnotationChecker

phaller commented:
When @TiarkRompf gives his green light it is OK to pull it into 2.10.x.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/246/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/953/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/953/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/246/

jsuereth commented:
We could really use a review on this!  @TiarkRompf is there someone else you can pass the ball to?

phaller commented:
@paulp offered to have a closer look at it.

heathermiller commented:
Bump- rumor has it that this is supposed to go into 2.10. Rumor also has it that @paulp is to review it?

gkossakowski commented:
What's the status of it?

paulp commented:
I'm supposed to look at it; but so far I haven't been able to spin up on it.  I'll try to at least make an assessment.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Modes.scala:
```diff
@@ -86,6 +86,10 @@ trait Modes {
+   */
+  final val RETmode       = 0x20000
+
   final private val StickyModes   = EXPRmode | PATTERNmode | TYPEmode | ALTmode
```
should RETmode be sticky? (what if you have a block expression as the result of the return and somewhere in there is a CPS expr that needs to be rewritten?)  In any case, this could be addressed in a future pull request.

adriaanm commented:
LGTM in the sense that I agree it's a conservative fix -- it shouldn't adversely affect non-CPS code
I can't comment on the correctness of the CPS transform

gkossakowski commented:
Thanks guys. I'll keep it in mind. In case of any regression with CPS we'll just revert these changes.

gkossakowski commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/642/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1350/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/642/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1350/

paulp commented:
Thanks, sorry I didn't manage this one.  I still have reservations about adding another typing mode.  Maybe we could undertake a documentation effort regarding how all those modes are supposed to work and (especially) how they're supposed to interact with one another, and what invariants should exist about what mode flags are set simultaneously.
